Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/calculation/CheckInWarningMatcher.kt

```diff
@@ -0,0 +1,143 @@
+
+import androidx.annotation.VisibleForTesting
+import de.rki.coronawarnapp.eventregistration.checkins.CheckIn
+import de.rki.coronawarnapp.eventregistration.checkins.split.splitByMidnightUTC
+import de.rki.coronawarnapp.presencetracing.warning.storage.TraceWarningPackage
+import de.rki.coronawarnapp.server.protocols.internal.pt.TraceWarning
+import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
+import de.rki.coronawarnapp.util.toOkioByteString
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.withContext
+import org.joda.time.Instant
+import timber.log.Timber
+import java.lang.reflect.Modifier.PRIVATE
+import javax.inject.Inject
+import kotlin.coroutines.CoroutineContext
+
+class CheckInWarningMatcher @Inject constructor(
+    private val dispatcherProvider: DispatcherProvider
+) {
+    suspend fun process(
+        checkIns: List<CheckIn>,
+        warningPackages: List<TraceWarningPackage>
+    ): Result {
+        val splitCheckIns = checkIns.flatMap { it.splitByMidnightUTC() }
+
+        val matchLists: List<List<MatchesPerPackage>?> = createMatchingLaunchers(
+            splitCheckIns,
+            warningPackages,
+            dispatcherProvider.IO
+        ).awaitAll()
+
+        val successful = if (matchLists.contains(null)) {
+            Timber.e("Calculation partially failed.")
+            false
+        } else {
+            Timber.d("Matching was successful.")
+            true
+        }
+
+        return Result(
+            successful = successful,
+            processedPackages = matchLists.filterNotNull().flatten()
+        )
+    }
+
+    data class Result(
+        val successful: Boolean,
+        val processedPackages: Collection<MatchesPerPackage> = emptyList(),
+    )
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    internal suspend fun createMatchingLaunchers(
+        checkIns: List<CheckIn>,
+        warningPackages: List<TraceWarningPackage>,
+        coroutineContext: CoroutineContext
+    ): List<Deferred<List<MatchesPerPackage>?>> {
+
+        val launcher: CoroutineScope.(
+            List<CheckIn>,
+            List<TraceWarningPackage>
+        ) -> Deferred<List<MatchesPerPackage>?> = { list, packageChunk ->
+            async {
+                try {
+                    packageChunk.map {
+                        val overlaps = findMatches(list, it)
+                        Timber.d("%d overlaps for %s", overlaps.size, it.packageId)
+                        MatchesPerPackage(warningPackage = it, overlaps = overlaps)
+                    }
+                } catch (e: Throwable) {
+                    Timber.e(e, "Failed to process packages $packageChunk")
+                    null
+                }
+            }
+        }
+
+        // at most 4 parallel processes
+        val chunkSize = (checkIns.size / 4) + 1
```

Comment:
The IO dispatcher would allow for up to 64? threads and each comparison will read the cached protobuf from storage. Too many simultaneous file operations may starve the system on I/O operations, lagging the whole device if we running this while our app is in the background. Once we know how large the packages will actually be we could think about increasing this, but I don't think it will have a huge impact and anything from 2-4 is fine.